MediaTracker 1.0

bonukai opened the pull request:
Major rewrite of entire project, frontend and backend, not yet completed.

Fronted has been fully redesign, but the progress on this is just starting. I have mainly focused on the backend rewrite. I have used [trpc](https://github.com/trpc/trpc) for API routes and [zod](https://github.com/colinhacks/zod) for type checking. This allows for easier development and reduces type errors.

I have optimized the database, now queries are much faster and simpler, but they still allow for complex filters and sorting. Updating metadata, especially for TV shows, have been reworked as well. When metadata has changed for a tv season, and it cannot be updated (because user have marked those episodes as seen), only this season update will be skipped, rest of metadata will be updated.  

## Breaking changes
- entire REST API has been rewritten, with the exception of the routes used by plugins
- translations of backend logs have been removed
- all environmental variables has been removed. Configuration can be set by web ui or cli 
- environmental variables in Docker has been removed or changed

## Migration instructions
For those using SQLite (default) there are no extra steps, just update docker image. 

If you are using a Postgres database, change environmental variable `DATABASE_CLIENT` from `pg` to `PostgreSQL` and use `DATABASE_CONNECTION_STRING` in form of `postgres://user:password@host:port/database` 

## New features:
- [x] #560 
- [x] #558 
- [x] #553   
- [x] #530 
- [x] #499  
- [x] #436 
- [x] #428 
- [x] #374
- [x] #171 
- [x] #229 
- [x] #156 
- [x] #126 
- [x] #118 
- [x] #116  
- [x] Fully customizable home screen
- [x] #115 
- [x] #157  
- [x] #570 

## Fixes
- [x] #561 
- [ ] #568
- [x] #554  

## TODO:
- [ ] details page
- [ ] season page
- [ ] progress page
- [ ] progress
- [x] notifications
- [ ] Trakt.tv import
- [ ] statistics
- [ ] mobile version
- [ ] tests
- [x] Docker image
- [ ] bundle fonts

bonukai committed:
initial commit

bonukai committed:
remove unused import

bonukai committed:
add migration for depreciated sort by "rank"

bonukai committed:
update .gitignore

bonukai committed:
update dependencies

bonukai committed:
add README

bonukai committed:
Update license

bonukai committed:
update VScode settings

bonukai committed:
update package.json

bonukai committed:
add VScode workplace

bonukai committed:
update dependencies

bonukai committed:
add Dockerfile

bonukai committed:
update eslint config

bonukai committed:
remove eslint comments

bonukai committed:
remove unused hook

bonukai committed:
check for nullable property

bonukai committed:
remove non-null assertions

bonukai committed:
add option to use TvMaze for episode release dates

bonukai committed:
add TMDB disclaimer

code-factor committed:
[CodeFactor] Apply fixes

bonukai committed:
fix pagination where number of pages is less then 8

bonukai committed:
save and display prefix of application token

bonukai committed:
update tasks.json

bonukai committed:
change nonempty to min in zod schema

bonukai committed:
add lingui to the server

bonukai committed:
update dependencies

mtricht commented on client/src/App.tsx:
```diff
@@ -1,95 +1,26 @@
-import { DarkModeProvider } from 'src/hooks/darkMode';
+import { FC } from 'react';
+import { RouterProvider } from 'react-router-dom';
+import { router } from './router';
```
```suggestion
import { router } from './Router';
```

mtricht commented on client/src/components/MediaItemViews.tsx:
```diff
@@ -0,0 +1,511 @@
+import { Plural, Trans } from '@lingui/macro';
+
+import { Button } from './Button';
+import { RelativeTime } from './Date';
```
```suggestion
import { RelativeTime } from './date';
```
Or rename to `date.tsx` to `Date.tsx`

bonukai replied:
I have renamed `date.tsx` to `Date.tsx`

mtricht reviewed (changes requested):
Looking great! Can't wait for the release, Two minor comments that I needed to change to get this running on my machine.
And three questions:
- You use prettier for backend and ESlint for frontend, why not just one?
- Is dark mode on your todo list?
- This PR is getting rather large, might I suggest introducing a 1.0 branch and merging changes into there?

bonukai committed:
Rename date.tsx to Date.tsx

bonukai committed:
Update client/src/App.tsx

rename import from  './router' to  './Router'

Co-authored-by: Michael van Tricht <metricht@gmail.com>

bonukai commented:
> * You use prettier for backend and ESlint for frontend, why not just one?

Initially I was using prettier for formatting and later I have added ESlint for linting, but it make sense to use ESlint for also for formatting. First I have to fix errors that ESlint is showing on the backend :)

> * Is dark mode on your todo list?

Yes, dark mode is on my TODO list, but not on the initial release

> * This PR is getting rather large, might I suggest introducing a 1.0 branch and merging changes into there?

I wanted to do that, but I couldn't crate a pull request to merge into this branch, but now I have tried again, and obviously a change needs to be made on this branch, not on `main` 🤦

bonukai committed:
Merge branch 'wip' of github.com:bonukai/MediaTracker into wip

bonukai committed:
Update ConfigurationPage.tsx (#582)

rename "foo" variable to "countryOptions"

bonukai commented on client/src/pages/settings/ConfigurationPage.tsx:
```diff
@@ -0,0 +1,656 @@
+  { code: 'ZW', country: 'Zimbabwe' },
+] as const;
+
+const foo = [
```
renamed in #582

bonukai commented on Dockerfile:
```diff
@@ -1,102 +1,96 @@
-ENV NODE_ENV=production
-
-ENTRYPOINT  ["sh", "/docker/entrypoint.sh"]
+FROM alpine:3.18 as alpine
```
Which line are you referring to?

```Dockerfile
FROM alpine:3.18 as alpine
```

Building of the Docker image is done `node:21-alpine3.18`, but the finished image is based on `alpine:3.18`, to reduce size. Those two need to be on the same alpine version, and for simplicity they are both defined at the top of the file. `alpine:3.18` is latter references as `alpine`

mtricht replied:
I see, my bad

bonukai committed:
remove version from client/package.json (#587)

bonukai committed:
fix: show log msg "Running migrations" before running migrations, not after (#589)

bonukai commented on client/package.json:
```diff
@@ -1,97 +1,44 @@
+  "name": "mediatracker",
   "private": true,
-  "proxy": "http://localhost:7481",
+  "version": "0.0.0",
```
Removed version in #587

bonukai committed:
fix: add migration to rename "my-rating" to "user-rating" in list sortBy field (#590)

bonukai committed:
fix: display message, if there are no migrations to run (#591)

bonukai committed:
add Postgres support to Docker (#594)

* update README.md

* build CLI args from env variables

bonukai committed:
add "demo" mode (#592)

* add "disabled" option to Button and PasswordInput

* add "DEMO" option to configuration

* add "demoMode" to configurationRouter

* disable password change in demo mode

* add Postgres support to Docker (#594)

* update README.md

* build CLI args from env variables

* add "disabled" option to Button and PasswordInput

* add "DEMO" option to configuration

* add "demoMode" to configurationRouter

* disable password change in demo mode

* add support for Dockerfile

bonukai committed:
rename "Configuration" to "Server configuration" (#595)

bonukai committed:
add extension to Router import (#596)

bonukai committed:
update audible type schema

bonukai committed:
small improvements to Checkbox

bonukai committed:
update types

bonukai committed:
format file

bonukai committed:
coerce numberOfPages

bonukai committed:
typing fixes and improvements (#606)

bonukai committed:
rename logo file

bonukai committed:
typing refactoring (#607)

bonukai committed:
Notifications (#608)

* implement notifications for new releases

* update translations

bonukai committed:
delete duplicate posters (#609)

* delete all posters - see #586

* delete only duplicates, instead of all files

* fix path bug

bonukai committed:
fix bug in notifications migration (#610)

bonukai committed:
fix: handle edge cases in migration (#611)

bonukai committed:
remove "totalRuntime" from list response (#612)

bonukai committed:
update dependencies (#613)

bonukai committed:
update translation files (#614)

bonukai committed:
Convert all release dates to UTC (#617)

bonukai committed:
fix typo

bonukai committed:
add error handling to API router (#620)

bonukai committed:
Fix browser "freeze" when downloading new images (#621)

bonukai committed:
update "server-configuration" link

bonukai committed:
fix bug in updateImage

bonukai committed:
update .gitignore

bonukai committed:
Merge pull request #622 from bonukai/patch

Patch

Commit:
remove redundant @types/sharp

bonukai committed:
Update README.md

Commit:
create database directory if not exists, for sqlite

Commit:
Merge branch 'wip' of https://github.com/bonukai/MediaTracker into wip

Commit:
update link to configuration page

bonukai committed:
fix: add error handling to notification functions

bonukai committed:
fix: further improvements on posters

bonukai committed:
Merge pull request #625 from bonukai/img

fix: further improvements on posters

bonukai committed:
parse all release dates without time as midnight in local time

bonukai committed:
Merge pull request #626 from bonukai/date

parse all release dates without time as midnight in local time

bonukai committed:
fix: use episode runtime from TMDB

bonukai committed:
update typescript

bonukai committed:
store runtime in milliseconds

bonukai committed:
Merge pull request #629 from bonukai/runtime

store runtime in milliseconds

bonukai committed:
update typescript

bonukai committed:
schedule updates last and upcoming episodes

bonukai committed:
Merge pull request #631 from bonukai/lastAndUpcomingEpisode

schedule updates last and upcoming episodes

ramebd committed:
fix option parsing to allow use of postgresql

ramebd committed:
fix the following error when picking the max seen date per mediaitem.

ERROR: error: select "date", "mediaItemId", max("date") as "date" from "seen" where "userId" = $1 and "mediaItemId" in ($2, $3) group by "mediaItemId" - column "seen.date" must appear in the GROUP BY clause or be used in an aggregate function

ramebd committed:
fix issue with numeric values being returned as strings

this broke mediaItemModelSchema because mediaItem.tmdbRating is defined as a decimal(8,2) value, but the database query returned a string due to no implicit type conversion from decimal to float in typescript

ramebd committed:
remove sortQuery when doing a simple count(*)

ramebd committed:
remove select * when doing a count(*)

bonukai committed:
update option parsing

bonukai committed:
rename migration and update it to retain data

bonukai committed:
fix bug, where a directory was created with the sqlite database if it does not exits

bonukai committed:
make id column in seenEpisodesCount auto increment

bonukai committed:
simplify pg types parsing

bonukai committed:
remove mysql client

bonukai committed:
Merge branch 'wip' into wip

bonukai committed:
clear all upcoming and last aired episodes when performing system wide update

bonukai committed:
fix add list popup not showing

bonukai committed:
fix upcoming and last aired episode

ramebd committed:
Merge pull request #2 from bonukai/wip

Wip catchup

bonukai committed:
fix justWatchProviders and seen query

bonukai committed:
Merge branch 'wip' of github.com:ramebd/MediaTracker into wip

bonukai committed:
Merge pull request #644 from ramebd/wip

WIP updates to allow postgres database connectivity

bonukai committed:
update dependencies

bonukai committed:
last and next airing

bonukai committed:
Merge pull request #646 from bonukai/patch

last and next airing

bonukai committed:
configure vitest

bonukai committed:
database transaction fix

bonukai committed:
optimize list items seen filter

bonukai committed:
Merge pull request #647 from bonukai/patch

optimize list items seen filter

bonukai committed:
implement #638

bonukai committed:
add github actions

bonukai committed:
Update tests.yml

bonukai committed:
update dependencies

bonukai committed:
Update tests.yml

bonukai committed:
update Dockerfile

bonukai committed:
update esbuild.js script

bonukai committed:
Update tests.yml

bonukai committed:
ryot import

bonukai committed:
Merge pull request #658 from bonukai/ryot_import

implement Ryot import

bonukai committed:
fix next/last airing date sort order on SQLite

bonukai committed:
add scroll to list selection

bonukai committed:
check for upcoming airing only for the next 24 hours

bonukai committed:
update sessionKey schema

vgambier commented:
Is there anything I can do to help with this rewrite?

makefu commented:
i'd also love to see this project ascend to version 1.0 :) Cheers!
